proposals/0000-type-functions.rst

```diff
@@ -0,0 +1,309 @@
+    p1 ::                 isDigit '0' :~: True
+    p1 :: forall isDigit. isDigit '0' :~: True
+
+  And this is not what we want. When we write ``isDigit``, we mean the
+  term-level function from ``Data.Char``, not an implicitly quantified type
+  variable. This restriction could be addressed by changing implicit
+  quantification rules, which we leave out of scope of this proposal.
+
+* The function cannot be used in the module in which it is defined because we
+  need an object file (``.o``) to link to. This is the same limitation as we
+  have in Template Haskell, since we are reusing the underlying mechanism.
+
+* The function must be fully instantiated and applied to all inputs.
+  Furthermore, inputs and the output must be of **promotable** types. The set
+  of promotable types is implementation-defined, but it's going to include at
+  least ``Char``, ``Bool``, ``Natural``, ``String``, ``Maybe a``,
+  ``Either a b``, ``[a]``, ``(a, b)``, and so on.
+
+Even with those limitations in place, we expect the feature to be of great
+utility. Consider this excerpt from the documentation of the ``servant``
+library used to derive webservices from a declarative specification::
+
+  >>> -- GET /books/:isbn
+  >>> type MyApi = "books" :> Capture "isbn" Text :> Get '[JSON] Book
+
+You'll notice that the example is first given using the intuitive notation
+``GET /books/:isbn``, and then using the ``servant``-specific syntax ``:>``,
+``Capture``, and so on.
+
+Instead, we could imagine defining the endpoint using the intuitive notation
+only::
+
+  type MyApi = Endpoint "GET /books/:isbn" (Text -> Book) '[JSON]
+
+The only problem is that we'd need a type family ``Endpoint`` that could parse
+the intuitive specification, but type-level parsing using type families is
+difficult and slows down the compiler.
+
+However, if we could write the parser at the term level, i.e. introduce a
+helper function ``parseEndpoint :: String -> [Either String String]``, and then
+use it to define ``Endpoint``, then this approach would be viable.
+
+We could define all sorts of type-level parsers, from SQL queries to CSS color
+notation, opening up new possibilities for type-level programming.
+
+Proposed Change Specification
+-----------------------------
+
+* Extend the type syntax with qualified lowercase names ``M.fn``.
+
+* Kind checking the application of a term to zero or more type arguments,
+  ``M.fn t0 t1 ... tn``, is done as follows:
+
+  1. Check that ``M.fn`` is defined in another module that can be dynamically
+     loaded, otherwise report a type error.
+  2. Check that the types of arguments ``t0 t1 ... tn`` and the result type are
+     monomorphic and **promotable** (defined below), otherwise report a type
+     error.
+  3. Instantiate and solve constraints, e.g. ``Prelude.sum [2,3,5]`` is
+     desugared into ``Prelude.sum @Natural $dNumNatural [2,3,5]``.
+     If there's no valid instantiation or constraints cannot be solved, report
+     a type error.
```

Comment:
> Which instances are in scope for this constraint-solving process? Given that this happens during kind-checking, presumably instances in the current module may not have been type-checked yet?

Initially, that was true, because I had been implementing the variant with a new sort of type family(it was close to the third case in the alternatives section). 
And yes, during the kind checking of a family declaration instances from the same module were unavailable. I've checked it today. But now I am going to do both kind and type checking during rewriting. And I will check if the domestic instances are available at the moment. 
Thanks for the question.      

> Also, it sounds like constraints will (still) not be allowed in kinds? So presumably the following will be rejected during kind-checking because the application of `sum` is polymorphic in the type of the list?
> 
> ```haskell
> test :: Prelude.sum xs ~ 0 => ...
> ```

Yes, It would be rejected. Another option is just to give up here and to return an unevaluated application and therefore unsolved constraint because it will be done even with a specified kind of `xs`. But such a  will lead to an acceptance of potentially incorrect declarations. So we are going to support polymorphism at the type level in declarations, but not at the kind level: 
```haskell  
test :: Prelude.sum xs ~ 0 => ... -- not ok. Potentially incorrect declaration.  
``` 
```haskell 
test :: forall (xs :: [Nat]). (Prelude.sum xs ~ 0) => ... - ok 
```

> What about
> 
> ```haskell
> test2 :: (xs ~ [1], Prelude.sum xs ~ 0) => ...
> test3 :: (Prelude.sum xs ~ 0, xs ~ [1]) => ...
> ```
> 
> where depending on the order of kind-checking we might or might not know the final kind of `xs` when we look at the application of `sum`?

Internally we would treat `Prelude.sum` in this place as a special case of a type family. We are going to both check the kind and reduce it while rewriting the application and do it using a slightly similar approach: 
 1. Check if arguments contain type variables or stuck type family applications.
 2. If they don't, then check the kind and compile a function, apply it, etc. If they do, then run rewriter on such arguments. 
 3. If the rewritten arguments are ok, then check the kind and run the reduction process. If not then check if kinds of arguments could be specified and if they satisfy the type of the function. 
     1. If it is true, then just return the unevaluated application 
     2. If it's not, then throw an error 
A similar approach works well with type families now and I think it would definitely work with type functions.

Reply:
>```Haskell 
>test :: Prelude.sum xs ~ 0
>```
I am thinking now about the possibility to write monomorphic functions with inferring kind of an argument. like `test :: M.sumNat xs ~ 0 => ...`

Reply:
I'm afraid I don't really follow these comments. I think the proposal really needs a clear specification of how kind- and type-checking will work for these type functions. Deferring any kind checking for them until rewriting seems potentially troublesome to me, because the current rewriter assumes the type expressions being rewritten are already kind-correct.

Reply:
> Deferring any kind checking for them until rewriting seems potentially troublesome to me, because the current rewriter assumes the type expressions being rewritten are already kind-correct.

Yes, but I am going to try because now during kind checking we have no domestic instances and an application of polymorphic function fails. I've checked it in my old prototype with type families after you question.

Reply:
> I would suggest the proposal should describe the ideal ultimate state of the feature, not constrained by the prototype.

Ok, but how should I include something that I want to but still looking for a way to implement?

Reply:
I think it's best to separate concerns about *specification* -- what is the language feature you're trying to design? -- from *implementation* -- how will you get GHC to bend to your will? My recommendation is to describe here the feature you'd like, regardless of implementation. Then, perhaps in a separate part of the proposal, introduce the fact that some aspect of the design seems hard to implement. Perhaps we can all brainstorm together to figure out a clever way to implement the idea.